### PR TITLE
Annotate cms::cuda::HostAllocator::allocate() as thread safe

### PR DESCRIPTION
**Pull request by makortel**
#### PR description:

This should silence 80 static analyzer warnings that @jpata reported in https://mattermost.web.cern.ch/cms-o-and-c/pl/6efrf7q4s7yi8k3oaamjhwdr3a

#### PR validation:

Code compiles.

### CODE REVIEW

**Comment by makortel** [comment]
@cmsbuild, please test

File: HeterogeneousCore/CUDAUtilities/interface/HostAllocator.h [hunk 1]
```diff
@@ -28,7 +30,7 @@ namespace cms {
         using other = HostAllocator<U, FLAGS>;
       };
 
-      T* allocate(std::size_t n) const __attribute__((warn_unused_result)) __attribute__((malloc))
+      T* allocate CMS_THREAD_SAFE(std::size_t n) const __attribute__((warn_unused_result)) __attribute__((malloc))
```
Comment (fwyzard) [inline HeterogeneousCore/CUDAUtilities/interface/HostAllocator.h:33]: is it the right place, between `allocate` and `(std::size_t n)` ?

Comment (makortel) [reply HeterogeneousCore/CUDAUtilities/interface/HostAllocator.h:33]: That appears to be a legitimate place for C++ attributes, honestly I was following an example in
https://github.com/cms-sw/cmssw/blob/8ffadd15b51612959296f749c1f0269947d3b3f5/DataFormats/TrackReco/interface/Track.h#L74-L76

On a further thought I agree the placement looks a bit weird. @gartung, could you comment how picky the static analyzer is for the placement of `[[cms::thread_safe]]` attribute? Could it be placed e.g. before the function declaration? (that we seem to do with `CMS_SA_ALLOW` here
https://github.com/cms-sw/cmssw/blob/8ffadd15b51612959296f749c1f0269947d3b3f5/FWCore/Concurrency/interface/WaitingTaskHolder.h#L77
)

Comment (fwyzard) [reply HeterogeneousCore/CUDAUtilities/interface/HostAllocator.h:33]: as far as I know, both
```
CMS_THREAD_SAFE T* allocate(std::size_t n) const __attribute__((warn_unused_result)) __attribute__((malloc)) 
```
and
```
T* allocate(std::size_t n) const CMS_THREAD_SAFE __attribute__((warn_unused_result)) __attribute__((malloc)) 
```
(or after the other attributes) should be fine

Comment (fwyzard) [inline HeterogeneousCore/CUDAUtilities/interface/HostAllocator.h:33]:
```suggestion
      CMS_THREAD_SAFE T* allocate(std::size_t n) const __attribute__((warn_unused_result)) __attribute__((malloc))
```

Comment (gartung) [reply HeterogeneousCore/CUDAUtilities/interface/HostAllocator.h:33]: As I recall there was some trial and error on where to place the attribute to the tokenizer would process it and attach it to the correct statement.

Comment (makortel) [reply HeterogeneousCore/CUDAUtilities/interface/HostAllocator.h]: Let's try this

Comment (gartung) [reply HeterogeneousCore/CUDAUtilities/interface/HostAllocator.h]: clang-check -ast-dump -ast-dump-filter=function-name will show you the AST the tokenizer produces and you can see if the attribute is attached to the function.

**Comment by makortel** [comment]
Seems that this way works as well, the [static analyzer report](https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-33d981/13889/llvm-analysis/) is not complaining about const member function returning a pointer to non-const object.

**Comment by gartung** [comment]
I noticed a bug in the summary which claims there are ESR::get warnings in the static analyzer report but there are not. I figured out the bug in the script and added a fix to 
https://github.com/cms-sw/cms-bot/pull/1527

**Comment by makortel** [comment]
+heterogeneous

**Comment by silviodonato** [comment]
+1

**Pull request merged**